doc/smart-switch/high-availability/smart-switch-ha-hld.md

```diff
@@ -0,0 +1,2176 @@
+
+#### 4.1.1. Network topology for Flow HA
+
+Today, the customers' traffic is partitioned into ENIs. And for each ENI:
+
+- We will have multiple DPUs forming a HA set for storing the traffic flows. Today, each HA set will only have 2 DPUs acting as a pair.
+- All DPUs in a HA set will be placed within a single Tier 1 set, but each DPU in the HA set will be placed on different T1s to avoid switch level failure killing the whole HA set.
+- Within a HA set, only 1 DPU will be serving as active and making decisions, while the other DPU will be serving as standby, only accepting flows that being replicated from the active one. 
+- When the active DPU runs into issues, we will failover the active, make the standby the new active, and switch over the traffic to avoid further impact.
+
+<p align="center"><img src="./images/ha-network-topology.svg"></p>
+
+#### 4.1.2. HA set setup
+
+To provide HA functionality, we are going to use Active-Standby setup. In this setup, 1 DPU will act as active and making decisions, while the non-active node will act as a pure backup flow storage.
+
+- When packets land on active DPU and creates a new flow, the new flow will be replicated to the standby DPU inline.
+- When packets land on standby side, they will be tunneled to the active DPU by NPU (or DPU as fallback). 
+
+<p align="center"><img src="./images/ha-active-standby-setup.svg"></p>
+
+### 4.2. Data path setup
+
+Besides providing HA for flows, we also need to provide solutions to ensure data path works with HA considered:
+
+1. Ensures having enough redundant links for the traffic to reach its target HA set.
+2. Ensures having enough redundant links for the 2 DPUs in the HA set to talk to each other.
+
+And this can be addressed by the ECMP setup in our network, as the "[Network topology for Flow HA](#network-topology-for-flow-ha)" section shows above.
+
+
+#### 4.2.1. Upstream-to-Switch communication channel
+
+In HA setup, our upstream service will talk to our northbound interface for:
+
+- Programming SDN dash policies.
+- Initiate the planned operation, or unplanned operation which can be destructive.
+- Get HA states and receive the HA state reports.
+- Setting up the traffic forwarding tunnels.
+
+#### 4.2.2. NPU-to-DPU traffic forward tunnel
+
+This channel is used for tunneling the packet from all T1s that has the traffic forwarding rules to the active side in the HA pair, no matter if it is locally or remotely.
+
+There is a limitation on the traffic forwarding rules, it can only forward the traffic to either local DPU or remote DPU, but never both as an ECMP group. This is caused by the limitation of routes. Therefore, the tunnel will always only have 1 next hop at maximum.
+
+It works as below in high level:
+
+1. The tunneled packets ***MUST*** have an VxLan encap as the outer packet, with a specified NPU tunnel VNI number as identifier.
```

Comment:
From DASH perspective, I believe so. Limiting the tunnel to VxLan has 2 purposes - to simplify the system since we only need to handle 1 single encap, as well as achieving better data path availability because we can manipulate the source port in VxLan encap.